tests/dash/gnmi_utils.py

```diff
@@ -0,0 +1,377 @@
+    dut_command = "docker exec %s pkill telemetry" % (env.gnmi_container)
+    duthost.shell(dut_command, module_ignore_errors=True)
+    dut_command = "docker exec %s bash -c " % env.gnmi_container
+    dut_command += "\"/usr/bin/nohup /usr/sbin/telemetry -logtostderr --port %s " % port
+    dut_command += "--server_crt /etc/sonic/telemetry/%s " % (env.gnmi_server_cert)
+    dut_command += "--server_key /etc/sonic/telemetry/%s " % (env.gnmi_server_key)
+    dut_command += "--ca_crt /etc/sonic/telemetry/%s " % (env.gnmi_ca_cert)
+    if env.enable_zmq:
+        dut_command += " -zmq_address=tcp://127.0.0.1:8100 "
+    dut_command += "-gnmi_native_write=true -v=10 >/root/gnmi.log 2>&1 &\""
+    duthost.shell(dut_command)
+    time.sleep(env.gnmi_server_start_wait_time)
+
+
+def recover_gnmi_cert(localhost, duthost):
+    env = GNMIEnvironment(duthost)
+    localhost.shell("rm -rf "+env.work_dir, module_ignore_errors=True)
+    dut_command = "docker exec %s supervisorctl status %s" % (env.gnmi_container, env.gnmi_program)
+    output = duthost.command(dut_command, module_ignore_errors=True)['stdout'].strip()
+    if 'RUNNING' in output:
+        return
+    dut_command = "docker exec %s pkill telemetry" % (env.gnmi_container)
+    duthost.shell(dut_command, module_ignore_errors=True)
+    dut_command = "docker exec %s supervisorctl start %s" % (env.gnmi_container, env.gnmi_program)
+    duthost.shell(dut_command)
+    time.sleep(env.gnmi_server_start_wait_time)
+
+
+def gnmi_set(duthost, ptfhost, delete_list, update_list, replace_list):
+    env = GNMIEnvironment(duthost)
+    ip = duthost.mgmt_ip
+    port = env.gnmi_port
+    cmd = 'python2 /root/gnxi/gnmi_cli_py/py_gnmicli.py '
+    cmd += '--timeout 30 '
+    cmd += '-t %s -p %u ' % (ip, port)
+    cmd += '-xo sonic-db '
+    cmd += '-rcert /root/%s ' % (env.gnmi_ca_cert)
+    cmd += '-pkey /root/%s ' % (env.gnmi_client_key)
+    cmd += '-cchain /root/%s ' % (env.gnmi_client_cert)
+    cmd += '-m set-update '
+    xpath = ''
+    xvalue = ''
+    for path in delete_list:
+        path = path.replace('sonic-db:', '')
+        xpath += ' ' + path
+        xvalue += ' ""'
+    for update in update_list:
+        update = update.replace('sonic-db:', '')
+        result = update.rsplit(':', 1)
+        xpath += ' ' + result[0]
+        xvalue += ' ' + result[1]
+    for replace in replace_list:
+        replace = replace.replace('sonic-db:', '')
+        result = replace.rsplit(':', 1)
+        xpath += ' ' + result[0]
+        if '#' in result[1]:
+            xvalue += ' ""'
+        else:
+            xvalue += ' ' + result[1]
+    cmd += '--xpath ' + xpath
+    cmd += ' '
+    cmd += '--value ' + xvalue
+    output = ptfhost.shell(cmd, module_ignore_errors=True)
+    error = "GRPC error\n"
+    if error in output['stdout']:
+        result = output['stdout'].split(error, 1)
+        return -1, result[1]
+    if output['stderr']:
+        return -1, output['stderr']
```

Comment:
Instead of returning the error code and error message, I suggest throwing an exception. It is a Python way to implement error handling and simplifies code a lot.

Reply:
updated